### PR TITLE
New module: `bamtools/convert`

### PR DESCRIPTION
**Pull request by sguizard**
## PR checklist
A new module for bamtools convert. It can convert bam files into: bed, fasta, fastq, json, pileup, sam, yaml

- [x] This comment contains a description of changes (with reason).
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/bamtools/convert/main.nf
Comment (chris-cheshire) [inline modules/bamtools/convert/main.nf:14]: Could change the output var name - this would be written in a pipeline as BAMTOOLS_CONVERT.out.out

File: tests/modules/bamtools/convert/main.nf
Comment (chris-cheshire) [inline tests/modules/bamtools/convert/main.nf:1]: You dont have any tests which cover your error execution path

File: tests/modules/bamtools/convert/main.nf
Comment (chris-cheshire) [inline tests/modules/bamtools/convert/main.nf:2]: You also only have one test - I would test for all supported file conversions

**Review by chris-cheshire — CHANGES_REQUESTED** [review]

File: modules/bamtools/convert/main.nf
Comment (sguizard) [reply modules/bamtools/convert/main.nf]: Of course, I changed it to `data`

File: tests/modules/bamtools/convert/main.nf
Comment (sguizard) [reply tests/modules/bamtools/convert/main.nf:2]: Done, I added a test for each supported format.

File: tests/modules/bamtools/convert/main.nf
Comment (sguizard) [reply tests/modules/bamtools/convert/main.nf:1]: Can you explain what you are expecting?

Comment (chris-cheshire) [reply tests/modules/bamtools/convert/main.nf:1]: You can configure to run tests that expect a failure, your testing coverage needs to cover your error state.

For example from - https://github.com/nf-core/cutandrun/blob/master/tests/test_samplesheet_check.yml

```
- name: test_samplesheet_check_noigg_neg
  command: nextflow run main.nf -profile docker,test --only_input true --igg_control true --input https://raw.githubusercontent.com/nf-core/test-datasets/cutandrun/samplesheet/unit_tests/no_igg.csv -c tests/config/nextflow.config
  tags:
    - samplesheet
    - samplesheet_no_igg
    - samplesheet_no_igg_neg
  exit_code: 1
```

File: tests/modules/bamtools/convert/main.nf
Comment (chris-cheshire) [reply tests/modules/bamtools/convert/main.nf:2]: Nice

**Review by grst — APPROVED** [review]
LGTM - I'm just wondering if it would make more sense to pass the format as an additional value channel rather than via `ext.args`? 

But both is possible - I'll leave that up to you!

File: tests/modules/bamtools/convert/main.nf
Comment (sguizard) [reply tests/modules/bamtools/convert/main.nf:1]: OK, I understand what you asking for. 
Presently, there is no guidelines for defining error tests.
I tried to add one test with an invalid extension, but pytest complains it do not find the versions.yml file. 
Any suggestion? 
```
(nf-core) ┬─[sguizard:~/W/D/g/n/modules]─[16:03:10]─[G:bamtools_convert↓9|✚3…1⚑1]
╰─>$ PROFILE=docker pytest --tag bamtools/convert --symlink --keep-workflow-wd
============================================================== test session starts ==============================================================
platform linux -- Python 3.9.7, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/sguizard/Work/Dev/github/nf-core/modules
plugins: workflow-1.6.0
collecting ... 
collected 626 items                                                                                                                             

bamtools convert test_bamtools_convert_invalid_ext:
	command:   nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_invalid_ext -c tests/config/nextflow.config
	directory: /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext
	stdout:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext/log.out
	stderr:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext/log.err
'bamtools convert test_bamtools_convert_invalid_ext' done.

bamtools convert test_bamtools_convert_bed:
	command:   nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_bed -c tests/config/nextflow.config
	directory: /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_bed
	stdout:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_bed/log.out
	stderr:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_bed/log.err
'bamtools convert test_bamtools_convert_bed' done.

bamtools convert test_bamtools_convert_fasta:
	command:   nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_fasta -c tests/config/nextflow.config
	directory: /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_fasta
	stdout:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_fasta/log.out
	stderr:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_fasta/log.err
'bamtools convert test_bamtools_convert_fasta' done.

bamtools convert test_bamtools_convert_fastq:
	command:   nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_fastq -c tests/config/nextflow.config
	directory: /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_fastq
	stdout:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_fastq/log.out
	stderr:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_fastq/log.err
'bamtools convert test_bamtools_convert_fastq' done.

bamtools convert test_bamtools_convert_json:
	command:   nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_json -c tests/config/nextflow.config
	directory: /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_json
	stdout:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_json/log.out
	stderr:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_json/log.err
'bamtools convert test_bamtools_convert_json' done.

bamtools convert test_bamtools_convert_pileup:
	command:   nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_pileup -c tests/config/nextflow.config
	directory: /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_pileup
	stdout:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_pileup/log.out
	stderr:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_pileup/log.err
'bamtools convert test_bamtools_convert_pileup' done.

bamtools convert test_bamtools_convert_sam:
	command:   nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_sam -c tests/config/nextflow.config
	directory: /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_sam
	stdout:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_sam/log.out
	stderr:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_sam/log.err
'bamtools convert test_bamtools_convert_sam' done.

bamtools convert test_bamtools_convert_yaml:
	command:   nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_yaml -c tests/config/nextflow.config
	directory: /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_yaml
	stdout:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_yaml/log.out
	stderr:    /tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_yaml/log.err
'bamtools convert test_bamtools_convert_yaml' done.

tests/test_versions_yml.py ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 17%]
sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 39%]
ssssssssssssssssssssssssssssssssssF.......sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 61%]
sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 83%]
sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss                                                                     [ 94%]
tests/modules/bamtools/convert/test.yml ....................................                                                              [100%] Keeping temporary directories and logs. Use '--kwd' or '--keep-workflow-wd' to disable this behaviour.


=================================================================== FAILURES ====================================================================
_______________________________ test_ensure_valid_version_yml[bamtools convert test_bamtools_convert_invalid_ext] _______________________________

workflow_dir = PosixPath('/tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext')

    @pytest.mark.workflow(*_get_workflow_names())
    def test_ensure_valid_version_yml(workflow_dir):
        workflow_dir = Path(workflow_dir)
        software_name = workflow_dir.name.split("_")[0].lower()
        try:
            versions_yml_file = workflow_dir / f"output/{software_name}/versions.yml"
>           versions_yml = versions_yml_file.read_text()

tests/test_versions_yml.py:28: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = PosixPath('/tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext/output/bamtools/versions.yml')
encoding = None, errors = None

    def read_text(self, encoding=None, errors=None):
        """
        Open the file in text mode, read it, and close the file.
        """
>       with self.open(mode='r', encoding=encoding, errors=errors) as f:

../../../../Softwares/anaconda3/envs/nf-core/lib/python3.9/pathlib.py:1266: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = PosixPath('/tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext/output/bamtools/versions.yml'), mode = 'r'
buffering = -1, encoding = None, errors = None, newline = None

    def open(self, mode='r', buffering=-1, encoding=None,
             errors=None, newline=None):
        """
        Open the file pointed by this path and return a file object, as
        the built-in open() function does.
        """
>       return io.open(self, mode, buffering, encoding, errors, newline,
                       opener=self._opener)

../../../../Softwares/anaconda3/envs/nf-core/lib/python3.9/pathlib.py:1252: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = PosixPath('/tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext/output/bamtools/versions.yml')
name = '/tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext/output/bamtools/versions.yml', flags = 524288
mode = 438

    def _opener(self, name, flags, mode=0o666):
        # A stub for the opener argument to built-in open()
>       return self._accessor.open(self, flags, mode)
E       FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext/output/bamtools/versions.yml'

../../../../Softwares/anaconda3/envs/nf-core/lib/python3.9/pathlib.py:1120: FileNotFoundError

During handling of the above exception, another exception occurred:

workflow_dir = PosixPath('/tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext')

    @pytest.mark.workflow(*_get_workflow_names())
    def test_ensure_valid_version_yml(workflow_dir):
        workflow_dir = Path(workflow_dir)
        software_name = workflow_dir.name.split("_")[0].lower()
        try:
            versions_yml_file = workflow_dir / f"output/{software_name}/versions.yml"
            versions_yml = versions_yml_file.read_text()
        except FileNotFoundError:
>           raise AssertionError(
                dedent(
                    f"""\
                    `versions.yml` not found in the output directory.
                    Expected path: `{versions_yml_file}`
    
                    This can have multiple reasons:
                    * The test-workflow failed before a `versions.yml` could be generated.
                    * The workflow name in `test.yml` does not start with the tool name.
                    """
                )
            )
E           AssertionError: `versions.yml` not found in the output directory.
E           Expected path: `/tmp/pytest_workflow_src0wgf8/bamtools_convert_test_bamtools_convert_invalid_ext/output/bamtools/versions.yml`
E           
E           This can have multiple reasons:
E           * The test-workflow failed before a `versions.yml` could be generated.
E           * The workflow name in `test.yml` does not start with the tool name.

tests/test_versions_yml.py:30: AssertionError
=============================================================== warnings summary ================================================================
../../../../Softwares/anaconda3/envs/nf-core/lib/python3.9/site-packages/pytest_workflow/plugin.py:391
  /home/sguizard/Work/Softwares/anaconda3/envs/nf-core/lib/python3.9/site-packages/pytest_workflow/plugin.py:391: UserWarning: .git dir detected: /home/sguizard/Work/Dev/github/nf-core/modules/.git. pytest-workflow will copy the entire .git directory and all files ignored by git. It is recommended to use the --git-aware option.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================ short test summary info ============================================================
FAILED tests/test_versions_yml.py::test_ensure_valid_version_yml[bamtools convert test_bamtools_convert_invalid_ext] - AssertionError: `versio...
============================================= 1 failed, 43 passed, 582 skipped, 1 warning in 49.79s =============================================
```

Comment (chris-cheshire) [reply tests/modules/bamtools/convert/main.nf:1]: Hey, you are using the old test running syntax, its now ` NF_CORE_MODULES_TEST=1 PROFILE=docker pytest --tag bamtools/convert`

Comment (sguizard) [reply tests/modules/bamtools/convert/main.nf:1]: Thanks for the updated command line. 
Unfortunately, I've got the same exact error with it.

Comment (chris-cheshire) [reply tests/modules/bamtools/convert/main.nf:1]: I also cant get the version output checking to run even with new syntax - I'm looking to see if this is a bug with the testing framework

**Comment by sguizard** [comment]
> LGTM - I'm just wondering if it would make more sense to pass the format as an additional value channel rather than via `ext.args`?
> 
> But both is possible - I'll leave that up to you!

Thanks for the review. 
For the format parameter, I tried to stick as much as possible to guideline (one channel = one input file). 
I can create a channel if you think it would improve the usage/readability of the code.

File: tests/modules/bamtools/convert/main.nf
Comment (chris-cheshire) [reply tests/modules/bamtools/convert/main.nf:1]: Okay I have some code suggestions for you, but there is a bug that needs to be fixed in the module testing code so that it does not check for a versions file if its checking for an error - I am raising a separate PR for this

```
- name: bamtools convert test_bamtools_convert_ext_error
  command: nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_ext_error -c tests/config/nextflow.config
  tags:
    - bamtools
    - bamtools/convert
  exit_code: 1

- name: bamtools convert test_bamtools_convert_noext_error
  command: nextflow run tests/modules/bamtools/convert -entry test_bamtools_convert_noext_error -c tests/config/nextflow.config
  tags:
    - bamtools
    - bamtools/convert
  exit_code: 1
```

```
    withName: BAMTOOLS_CONVERT_EXT_ERROR {
        ext.args = "-format testext"
    }

    withName: BAMTOOLS_CONVERT_NOEXT_ERROR {
        ext.args = ""
    }
```

```
workflow test_bamtools_convert_ext_error {

    input = [
        [ id:'test', single_end:false ], // meta map
        file(params.test_data['sarscov2']['illumina']['test_paired_end_bam'], checkIfExists: true)
    ]

    BAMTOOLS_CONVERT_EXT_ERROR ( input )
}

workflow test_bamtools_convert_noext_error {

    input = [
        [ id:'test', single_end:false ], // meta map
        file(params.test_data['sarscov2']['illumina']['test_paired_end_bam'], checkIfExists: true)
    ]

    BAMTOOLS_CONVERT_NOEXT_ERROR ( input )
}
```

Comment (sguizard) [reply tests/modules/bamtools/convert/main.nf:1]: OK, Thanks for your help. 
I pushed the code with tests.

**Comment by chris-cheshire** [comment]
Okay I got the fix pushed through quick so we can now use negative testing

File: tests/modules/bamtools/convert/test.yml
Comment (chris-cheshire) [inline tests/modules/bamtools/convert/test.yml:14]: Annoyingly - you can not put in custom tags - I have also wanted to do this before

**Review by chris-cheshire — CHANGES_REQUESTED** [review]

File: tests/modules/bamtools/convert/test.yml
Comment (sguizard) [reply tests/modules/bamtools/convert/test.yml]: 😄 No problems, I removed those. 
The tests passed locally. Let's on github.

**Review by chris-cheshire — APPROVED** [review]

**Comment by sguizard** [comment]
Thanks for the review!

**Pull request merged**
